File: huffman_coding_java/src/Decoder.java

import java.io.IOException;
import java.io.InputStream;


public class Decoder {

    private int currentByte;
    private int bitsLeft;
    private InputStream inputStream;
    private Node root;
    private StringBuilder stringBuilder;



    public Decoder(Node root, InputStream inputStream) {
        this.root = root;
        this.inputStream = inputStream;
        stringBuilder = new StringBuilder();
    }



    public String decodeFile() throws IOException {

        while(true) {
            char symbol = readNextSymbol();
            write(symbol);
            if(symbol == '□') break;
        }

        return stringBuilder.toString();

    }



    private char readNextSymbol() throws IOException {
        Node currentNode = root;
        while(true) {
            Node next = new Node();
            int dir = readBit();
            if(dir == -1) return 256;

            if(dir == 0) next = currentNode.left;
            if(dir == 1) next = currentNode.right;

            if(next.hasCharacter()) return next.c;
            else currentNode = next;
        }

    }



    private int readBit() throws IOException {

        if (currentByte == -1) return -1;

        if (bitsLeft == 0) {
            currentByte = inputStream.read();
            if (currentByte == -1) return -1;
            bitsLeft = 8;
        }

        if (bitsLeft <= 0)
            throw new AssertionError();

        bitsLeft--;
        return (currentByte >>> bitsLeft) & 1;

    }


    private void write(char symbol) {
        stringBuilder.append(symbol);
    }


    public String getContent() {
        return stringBuilder.toString();
    }
}

File: huffman_coding_scala/src/Decoder.scala
import java.io.FileInputStream

import scala.collection.mutable

class Decoder (var inputFileName: String = "", root: Node, var debug: Boolean = false) {

  var inputStream = new FileInputStream(inputFileName)
  var currentByte: Int = 0
  var bitsLeft: Int = 0

  var fileContent: String = ""
  def getContent: String = fileContent

  /**
    * Decodes the given file into a string builder, and flushes the builder upon the
    * last character of the file.
    */
  def decodeFile() {

    var sb: StringBuilder = new mutable.StringBuilder()
    while(true) {
      val symbol = nextSymbol().getOrElse({
        fileContent = sb.toString()
        if(debug) {
          println(fileContent)
        }
        return
      })
      sb += symbol
    }
  }

  /**
    * Reads the next char symbol from the file. This is done by
    * navigating the Code Tree for every bit in the file.
    * @return
    */
  def nextSymbol(): Option[Char] = {

    var currentNode = root
    while(true) {

      currentNode match {

        case internal: InternalNode => {
          var next: Node = internal.left
          val direction = readBit()
          if (direction == 1) next = internal.right
          currentNode = next
        }

        case leaf: LeafNode => return if(leaf.c != '□') Option(leaf.c) else None
      }
    }
    None
  }

  /**
    * Reads a single bit from the file by isolating the LSB in the
    * current byte.
    * @return
    */
  def readBit(): Int = {

    if(currentByte == -1) return -1

    if(bitsLeft == 0) {
      currentByte = inputStream.read()
      if(currentByte == -1) return -1
      bitsLeft = 8
    }

    bitsLeft -= 1
    (currentByte >>> bitsLeft) & 1
  }
}
